src/order/circular.lean

```diff
@@ -0,0 +1,389 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.set.basic
+
+/-!
+# Circular order hierarchy
+
+This file defines circular preorders, circular partial orders and circular orders.
+
+## Hierarchy
+
+* A `circular_order` is a ternary betweenness relation `btw : α → α → α → Prop`, which is
+  - reflexive: `btw a a a`
+  - cyclic: `btw a b c → btw b c a`
+  - antisymmetric: `btw a b c → btw c b a → a = b ∨ b = c ∨ c = a`
+  - total: `btw a b c ∨ btw c b a`
+  along with a strict betweenness relation `sbtw : α → α → α → Prop` which respects
+  `sbtw a b c ↔ btw a b c ∧ ¬ btw c b a`, analogously to how `<` and `≤` are related, and is
+  - transitive: `sbtw a b c → sbtw b d c → sbtw a d c`.
+* A `circular_partial_order` drops totality.
+* A `circular_preorder` further drops antisymmetry.
+
+The mind picture is that a circular order is a circle and `btw a b c` means that going around
+clockwise from `a` you reach `b` before `c` (`b` is between `a` and `c` is meaningless on an
+unoriented circle). A circular partial order is several, potentially intersecting, circles. A
+circular preorder is more of a mess.
+
+One can translate from usual orders to circular ones by "closing the necklace at infinity". See
+`has_le.to_has_btw` and `has_lt.to_has_sbtw`. Going the other way involves "cutting the necklace" or
+"rolling the necklace open".
+
+## Main definitions
+
+* `set.cIcc`: Closed-closed circular interval.
+* `set.cIoo`: Open-open circular interval.
+
+## Notes
+
+There's an unsolved diamond here. The instances `has_le α → has_btw (order_dual α)` and
+`has_lt α → has_sbtw (order_dual α)` can each be inferred in two ways:
+* `has_le α` → `has_btw α` → `has_btw (order_dual α)` vs
+  `has_le α` → `has_le (order_dual α)` → `has_btw (order_dual α)`
+* `has_lt α` → `has_sbtw α` → `has_sbtw (order_dual α)` vs
+  `has_lt α` → `has_lt (order_dual α)` → `has_sbtw (order_dual α)`
+The fields are propeq, but not defeq. It is temporarily fixed by turning the circularizing instances
+into definitions.
+
+## TODO
+
+Antisymmetry is quite weak in the sense that there's no way to discriminate which two points are
+equal. This prevents defining closed-open intervals `cIco` and `cIoc` in the neat `=`-less way. We
+currently haven't defined them at all.
+
+What's next is to define circular groups and provide instances for `zmod` and the usual circle group
```

Comment:
And even `roots_of_unity`, although that requires a choice of `complex.I : ℂ`, so maybe it shouldn't be an instance?

Reply:
Actually, that's pretty interesting. When can you circularly order `roots_of_unity`?

Reply:
They are isomorphic to `zmod n`, right?